### PR TITLE
Idp 718 creazione api per stato della transazione lato merchant

### PR DESCRIPTION
**Pull request by Giuseppe-LaManna**

### CODE REVIEW

File: src/main/java/it/gov/pagopa/payment/controller/TransactionController.java [hunk 1]
```diff
@@ -9,4 +9,6 @@ public interface TransactionController {
     @GetMapping("/{transactionId}")
     @ResponseStatus(code = HttpStatus.OK)
     TransactionInProgress getTransaction(@PathVariable String transactionId, @RequestHeader("x-user-id") String userId);
+
```
Comment (marco-di-lisa) [inline src/main/java/it/gov/pagopa/payment/controller/TransactionController.java:12]: If the only differences on this file are spaces, we can roll it back

File: src/main/java/it/gov/pagopa/payment/controller/TransactionControllerImpl.java [hunk 2]
```diff
@@ -19,4 +19,6 @@ public TransactionInProgress getTransaction(String transactionId, String userId)
         log.info("[GET_TRANSACTION] User {} requested to retrieve transaction {}", userId, transactionId);
         return transactionService.getTransaction(transactionId, userId);
     }
+
```
Comment (marco-di-lisa) [inline src/main/java/it/gov/pagopa/payment/controller/TransactionControllerImpl.java:22]: If the only differences on this file are spaces, we can roll it back

File: src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    private String operationType;
+    private Long amountCents;
+    private String amountCurrency;
+    private String mcc;
+    private String acquirerId;
+    private String merchantId;
+    private String initiativeId;
+    private Long rewardCents;
+    private List<String> rejectionReasons = new ArrayList<>();
+    private it.gov.pagopa.payment.enums.SyncTrxStatus status;
```
Comment (marco-di-lisa) [inline src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java:34]: It is better to import the class and use only the class name as the object type

File: src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java [hunk 4]
```diff
@@ -51,4 +63,12 @@ public TransactionResponse confirmPayment(String trxId, String merchantId) {
     return qrCodeConfirmationService.confirmPayment(trxId, merchantId);
   }
 
+  @Override
+  public SyncTrxStatusDTO getStatusTransaction(String transactionId, String merchantId, String acquirerId) {
```
Comment (marco-di-lisa) [inline src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java:67]: Let's move the business logic in a separated service as for the other APIs
In this class we will just call the external service

File: src/main/java/it/gov/pagopa/payment/service/TransactionService.java [hunk 5]
```diff
@@ -1,7 +1,9 @@
 package it.gov.pagopa.payment.service;
 
+
 import it.gov.pagopa.payment.model.TransactionInProgress;
 
 public interface TransactionService {
     TransactionInProgress getTransaction(String id, String userId);
+
```
Comment (marco-di-lisa) [inline src/main/java/it/gov/pagopa/payment/service/TransactionService.java:8]: If the only differences on this file are spaces, we can roll it back

File: src/main/java/it/gov/pagopa/payment/service/TransactionServiceImpl.java [hunk 6]
```diff
@@ -34,4 +36,6 @@ public TransactionInProgress getTransaction(String id, String userId) {
 
     return transactionInProgress;
   }
+
```
Comment (marco-di-lisa) [inline src/main/java/it/gov/pagopa/payment/service/TransactionServiceImpl.java:39]: If the only differences on this file are spaces, we can roll it back

File: src/test/java/it/gov/pagopa/payment/controller/TransactionControllerTest.java [hunk 7]
```diff
@@ -57,4 +54,6 @@ void getTransaction() throws Exception {
 
     Assertions.assertEquals(expectedTrx, trx);
   }
+
```
Comment (marco-di-lisa) [inline src/test/java/it/gov/pagopa/payment/controller/TransactionControllerTest.java:57]: If the only differences on this file are spaces, we can roll it back

**Review by marco-di-lisa — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java [hunk 8]
```diff
@@ -17,16 +23,22 @@ public class QRCodePaymentServiceImpl implements QRCodePaymentService {
   private final QRCodePreAuthService qrCodePreAuthService;
   private final QRCodeAuthPaymentService qrCodeAuthPaymentService;
   private final QRCodeConfirmationService qrCodeConfirmationService;
+  private final TransactionInProgressRepository transactionInProgressRepository;
+  private final TransactionInProgress2SyncTrxStatus transactionMapper;
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java:27]: I found transactionMapper too generic for a Facade class
```suggestion
  private final TransactionInProgress2SyncTrxStatus transaction2statusMapper;
```

File: src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java (hunk 4 above)
Comment (antonioT90) [reply src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java:75]: maybe just this time, only because there are very low code's rows, we could let it here :) 

When we will implement the CIEPaymentService, this entire class will be renamed into "BasePaymentService" (making a new QRCodePaymentService which will extend it), this way also the CIE flows will ineherit this logic

File: src/main/java/it/gov/pagopa/payment/service/TransactionService.java (hunk 5 above)
Comment (antonioT90) [reply src/main/java/it/gov/pagopa/payment/service/TransactionService.java:8]: Given the amount of these occurrences, I would describi why we are asking to you to avoid to commit these changes:

This way our review is full of class changed (actually 16) which not all of them has significant changes

File: src/test/java/it/gov/pagopa/payment/controller/QRCodePaymentControllerImplTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+
+        Assertions.assertNotNull(trx);
+        Assertions.assertEquals(trxStatus_2,trx);
+        Mockito.verify(qrCodePaymentService).getStatusTransaction(anyString(),anyString(),anyString());
+
+    }
+
+    @Test
+    void getStatusTransactionException() throws Exception {
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/controller/QRCodePaymentControllerImplTest.java:61]: let's use a more specific name
```suggestion
    void getStatusTransaction_NotFoundException() throws Exception {
```

File: src/test/java/it/gov/pagopa/payment/dto/mapper/TransactionInProgress2SyncTrxStatusTest.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package it.gov.pagopa.payment.dto.mapper;
+
+import org.junit.jupiter.api.Test;
+
+
+
+class TransactionInProgress2SyncTrxStatusTest {
+
+    @Test
+    void transactionInProgressMapper() {
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/dto/mapper/TransactionInProgress2SyncTrxStatusTest.java:10]: let's remember to check if there are not null fields (except what should be expected to be null)
it.gov.pagopa.payment.test.utils.TestUtils#checkNotNullFields

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    QRCodeCreationService qrCodeCreationService;
+    @Mock
+     QRCodePreAuthService qrCodePreAuthService;
+    @Mock
+    QRCodeAuthPaymentService qrCodeAuthPaymentService;
+    @Mock
+    QRCodeConfirmationService qrCodeConfirmationService;
+    @Mock
+    TransactionInProgress2SyncTrxStatus transactionMapper;
+    QRCodePaymentServiceImpl qrCodePaymentService;
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:41]: evenif it's a test, let's always respect the principle of least privileges

let's use suffix "Mock" to explicitly let know who would read it in the rest of the test, that these variables are mocks

let's separate using empty lines the variables by their "type" (mocks, utilities, tested service), this way we will simplify file reading

```suggestion
    private TransactionInProgressRepository transactionInProgressRepositoryMock;
    @Mock
    private QRCodeCreationService qrCodeCreationServiceMock;
    @Mock
     private QRCodePreAuthService qrCodePreAuthServiceMock;
    @Mock
    private QRCodeAuthPaymentService qrCodeAuthPaymentServiceMock;
    @Mock
    private QRCodeConfirmationService qrCodeConfirmationServiceMock;
    
    private final TransactionInProgress2SyncTrxStatus transactionMapper = new ....;
    
    private QRCodePaymentServiceImpl qrCodePaymentServiceMock;
```

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    @Mock
+    QRCodeCreationService qrCodeCreationService;
+    @Mock
+     QRCodePreAuthService qrCodePreAuthService;
+    @Mock
+    QRCodeAuthPaymentService qrCodeAuthPaymentService;
+    @Mock
+    QRCodeConfirmationService qrCodeConfirmationService;
+    @Mock
+    TransactionInProgress2SyncTrxStatus transactionMapper;
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:40]: A mapper is a so simple class (having a separate test which will guarantee for its right behavior) whose nature give us more benefits using its real implementation than its mock

We could trust its implementation (because there is a test that proof it) so could avoid to configure its mock, so let's just use the real implementation

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        //then
+        Assertions.assertNotNull(trx);
+        Assertions.assertEquals(trxStatus_1,trx);
+        Assertions.assertInstanceOf(SyncTrxStatusDTO.class,trx);
+        Mockito.verify(transactionInProgressRepository).findByIdAndMerchantIdAndAcquirerId(anyString(),anyString(),anyString());
+        Mockito.verify(transactionMapper).transactionInProgressMapper(any());
+    }
+
+    @Test
+    void getStatusTransactionException(){
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:73]:
```suggestion
    void getStatusTransaction_NotFoundException(){
```

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        String acquirerId="ACQUIRERID1";
+        TransactionInProgress transaction = TransactionInProgressFaker.mockInstance(1, SyncTrxStatus.IDENTIFIED);
+        doReturn(Optional.of(transaction)).when(transactionInProgressRepository).findByIdAndMerchantIdAndAcquirerId(transactionId,merchantId,acquirerId);
+        doReturn(trxStatus_1).when(transactionMapper).transactionInProgressMapper(transaction);
+        //when
+        SyncTrxStatusDTO trx= qrCodePaymentService.getStatusTransaction(transactionId,merchantId,acquirerId);
+        //then
+        Assertions.assertNotNull(trx);
+        Assertions.assertEquals(trxStatus_1,trx);
+        Assertions.assertInstanceOf(SyncTrxStatusDTO.class,trx);
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:67]: this is superfluous, it's obvious that this is its class!

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                transactionInProgressRepository,
+                transactionMapper);
+    }
+    @Test
+    void getStatusTransaction() {
+        //given
+        SyncTrxStatusDTO trxStatus_1= SyncTrxStatusFaker.mockInstance(1);
+        String transactionId="TRANSACTIONID1";
+        String merchantId="MERCHANTID1";
+        String acquirerId="ACQUIRERID1";
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:58]: these values are not really used, they are already stored inside TransactionStatus

.findByIdAndMerchantIdAndAcquirerId(trx.getTransactionId(),trx.getMerchantId(), trx.getAcquirerId());

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+        TransactionInProgress transaction = TransactionInProgressFaker.mockInstance(1, SyncTrxStatus.IDENTIFIED);
+        doReturn(Optional.of(transaction)).when(transactionInProgressRepository).findByIdAndMerchantIdAndAcquirerId(transactionId,merchantId,acquirerId);
+        doReturn(trxStatus_1).when(transactionMapper).transactionInProgressMapper(transaction);
+        //when
+        SyncTrxStatusDTO trx= qrCodePaymentService.getStatusTransaction(transactionId,merchantId,acquirerId);
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:63]: let's use "result" name, this way is more simple to let know in the then clause what should be tested

```suggestion
        SyncTrxStatusDTO result= qrCodePaymentService.getStatusTransaction(transactionId,merchantId,acquirerId);
```

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    @Test
+    void getStatusTransaction() {
+        //given
+        SyncTrxStatusDTO trxStatus_1= SyncTrxStatusFaker.mockInstance(1);
+        String transactionId="TRANSACTIONID1";
+        String merchantId="MERCHANTID1";
+        String acquirerId="ACQUIRERID1";
+        TransactionInProgress transaction = TransactionInProgressFaker.mockInstance(1, SyncTrxStatus.IDENTIFIED);
+        doReturn(Optional.of(transaction)).when(transactionInProgressRepository).findByIdAndMerchantIdAndAcquirerId(transactionId,merchantId,acquirerId);
+        doReturn(trxStatus_1).when(transactionMapper).transactionInProgressMapper(transaction);
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:61]: using the real implementation, this mock configuration will not be necessary

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        //when
+        SyncTrxStatusDTO trx= qrCodePaymentService.getStatusTransaction(transactionId,merchantId,acquirerId);
+        //then
+        Assertions.assertNotNull(trx);
+        Assertions.assertEquals(trxStatus_1,trx);
+        Assertions.assertInstanceOf(SyncTrxStatusDTO.class,trx);
+        Mockito.verify(transactionInProgressRepository).findByIdAndMerchantIdAndAcquirerId(anyString(),anyString(),anyString());
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:68]: The actual implementation of Mockito, configured as strict, will give you an error if the mock configuration is not used.
This let this "verify" useless

Rather, create a @AfterEach method where you will call
Mockito.verifyNotMoreInteraction
on all @Mocks

File: src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.List;
+
+public class SyncTrxStatusFaker {
+
+    public static SyncTrxStatusDTO mockInstance(Integer bias){
+        return mockInstanceBuilder(bias).build();
+    }
+    public static SyncTrxStatusDTO.SyncTrxStatusDTOBuilder mockInstanceBuilder(Integer bias){
+
+        return SyncTrxStatusDTO.builder()
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java:22]: Always due to the presence of the mapper's test, we could re-use the TransactionInProgress faker to create a valid instance of that object, then we could use the mapper to transform it into a SyncTrxStatus

This way, if there changes on the valid configuration, we should update just the TransactionInProgress faker

**Review by antonioT90 — CHANGES_REQUESTED** [review]

**Comment by antonioT90** [comment]
If you have time to spend, let's test also this API inside the it.gov.pagopa.payment.controller.BasePaymentControllerIntegrationTest
test, in order to check if we are able to call it inside at least 1 useCase obtaining the expected result (also the 404 when not allowed or not existend)

File: src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+@Builder
+public class SyncTrxStatusDTO {
+    private String id;
+    private String idTrxIssuer;
+    private String trxCode;
+    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
+    private OffsetDateTime trxDate;
+    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
+    private LocalDateTime authDate;
+    private String operationType;
```
Comment (sbellezz) [inline src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java:26]: The type of this field should be src/main/java/it/gov/pagopa/payment/enums/OperationType.java

Comment (marco-di-lisa) [reply src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java:26]: It depends on whether we want to return the operation type "00"/"01" or "CHARGE"/"REFUND".  The latter should require the enum. It is not mandatory, but it is safer

Comment (antonioT90) [reply src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java:26]: 00 or 01 are something inner to rewardCalculator, here let's use the enum ;)

Comment (Giuseppe-LaManna) [reply src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java:26]: This field in TransactionInProgress is a String. Do we want to replace them in an enum as well?

Comment (antonioT90) [reply src/main/java/it/gov/pagopa/payment/dto/qrcode/SyncTrxStatusDTO.java:26]: TransactionInProgress#operationType is something else!
we should mapping SyncTrxStatus#operationType with TrxInProgress#operationTypeTranscoded

If inside the mapper you are mapping TrxInProgress#operationType, that is an error not seen by us :)

File: src/main/java/it/gov/pagopa/payment/controller/QRCodePaymentController.java [hunk 21]
```diff
@@ -35,4 +30,8 @@ AuthPaymentDTO authPayment(@PathVariable("trxCode") String trxCode,
 
   @PutMapping("/merchant/{transactionId}/confirm")
   TransactionResponse confirmPayment(@PathVariable("transactionId") String trxId, @RequestHeader("x-merchant-id") String merchantId, @RequestHeader("x-acquirer-id") String acquirerId);
+
+  @GetMapping("/merchant/status/{transactionId}")
+  @ResponseStatus(code = HttpStatus.OK)
+  SyncTrxStatusDTO getStatusTransaction(@PathVariable String transactionId, @RequestHeader("x-merchant-id") String merchantId, @RequestHeader("x-acquirer-id") String acquirerId);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/controller/QRCodePaymentController.java:36]:
```suggestion
  SyncTrxStatusDTO getStatusTransaction(@PathVariable("transactionId") String transactionId, @RequestHeader("x-merchant-id") String merchantId, @RequestHeader("x-acquirer-id") String acquirerId);
```

File: src/main/java/it/gov/pagopa/payment/service/TransactionService.java [hunk 22]
```diff
@@ -1,5 +1,6 @@
 package it.gov.pagopa.payment.service;
 
+
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/service/TransactionService.java:3]: could you please revert this change?

File: src/main/java/it/gov/pagopa/payment/service/TransactionServiceImpl.java [hunk 23]
```diff
@@ -6,11 +6,13 @@
 import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Service;
 
+
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/service/TransactionServiceImpl.java:9]: could you please revert these changes?

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+    @Mock
+    TransactionInProgressRepository transactionInProgressRepositoryMock;
+    @Mock
+    QRCodeCreationService qrCodeCreationServiceMock;
+    @Mock
+     QRCodePreAuthService qrCodePreAuthServiceMock;
+    @Mock
+    QRCodeAuthPaymentService qrCodeAuthPaymentServiceMock;
+    @Mock
+    QRCodeConfirmationService qrCodeConfirmationServiceMock;
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:46]:
```suggestion
    private TransactionInProgressRepository transactionInProgressRepositoryMock;
    @Mock
    private QRCodeCreationService qrCodeCreationServiceMock;
    @Mock
     private QRCodePreAuthService qrCodePreAuthServiceMock;
    @Mock
    private QRCodeAuthPaymentService qrCodeAuthPaymentServiceMock;
    @Mock
    private QRCodeConfirmationService qrCodeConfirmationServiceMock;
```

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+    @Spy
+    private final TransactionInProgress2SyncTrxStatusMapper transactionMapper= new TransactionInProgress2SyncTrxStatusMapper();
+
+    QRCodePaymentServiceImpl qrCodePaymentServiceMock;
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:50]: this is not a mock
```suggestion
    private QRCodePaymentServiceImpl qrCodePaymentService;
```

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+        qrCodePaymentServiceMock = new QRCodePaymentServiceImpl(qrCodeCreationServiceMock,
+                qrCodePreAuthServiceMock,
+                qrCodeAuthPaymentServiceMock,
+                qrCodeConfirmationServiceMock,
+                transactionInProgressRepositoryMock,
+                transactionMapper);
+    }
+
+    @AfterEach
+    void tearDown() {
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:63]: why this name?
```suggestion
    void verifyNoMoreMockInteractions() {
```

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java (hunk 26 above)
Comment (Giuseppe-LaManna) [reply src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java:63]: the name of the tearDown method is the provided by junit, it is the same thing as the setUp

**Review by marco-di-lisa — APPROVED** [review]

File: src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java [hunk 27]
```diff
@@ -317,6 +327,28 @@ private TransactionInProgress checkIfStored(String trxId) {
             Assertions.assertFalse(transactionInProgressRepository.existsById(trxCreated.getId()));
         });
 
+        //useCase 6: trx Status given
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java:330]: it's too expensive to create a new useCase just to test the getStatus!
remove this!

simply inside method "checkTransactionStored" let's invoke the getStatusTransaction to check the 200 status and the matching content.
Then try invoking again using a mismatched merchantId, verifying the 404

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java [hunk 28]
```diff
@@ -132,8 +141,23 @@ private TransactionResponse createTrxSuccess(TransactionCreationRequest trxReque
...
+                    Assertions.assertEquals(trxCreated.getInitiativeId(), syncTrxStatusResult.getInitiativeId());
+                    Assertions.assertEquals(trxCreated.getMerchantId(), syncTrxStatusResult.getMerchantId());
+                    Assertions.assertEquals(trxCreated.getIdTrxIssuer(), syncTrxStatusResult.getIdTrxIssuer());
+                    Assertions.assertEquals(trxCreated.getTrxDate(), syncTrxStatusResult.getTrxDate());
+                    Assertions.assertEquals(trxCreated.getAmountCents(), syncTrxStatusResult.getAmountCents());
+                    Assertions.assertEquals(trxCreated.getAmountCurrency(), syncTrxStatusResult.getAmountCurrency());
+                    Assertions.assertEquals(trxCreated.getMcc(), syncTrxStatusResult.getMcc());
+                    Assertions.assertEquals(trxCreated.getAcquirerId(), syncTrxStatusResult.getAcquirerId());
+                    Assertions.assertEquals(trxCreated.getStatus(), syncTrxStatusResult.getStatus());
+                });
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java:159]: as done in row 262, use the mapper to assert the equals instead of repeating all the equality assertions (as alternative, you could re-use the assertions written in the mapper tests)!

Otherwise you should update these assertions each time you add a field to the mapped DTO!
And this is not the main purpose of this test class!

File: src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+import it.gov.pagopa.payment.enums.SyncTrxStatus;
+import it.gov.pagopa.payment.model.TransactionInProgress;
+import it.gov.pagopa.payment.utils.RewardConstants;
+
+import java.time.LocalDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+
+public class SyncTrxStatusFaker {
+    public static SyncTrxStatusDTO mockInstance(Integer bias){
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java:14]: when building a useful mock instance builder we are interested to have the more generic (and valid) form of the object (which probably will be the more often used form).

Here you are returning a precise use case, where the trx has been rejected due to the unexistance of the initiatives...

Probably you have done this because you are testing this particular useCase, but this could not be the more generic "mockInstance" that someone could be interested to retrieve!

I would let the generic "mockInstance" to simply return the "map" of the TransactionInProgress "mockInstance"

If the useCase implemented here is frequent, let's implement it inside a "mockNoInitiativeInstance"
which would call the "SyncTrxStatusFaker .mockInstance" setting the fields required to change

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java (hunk 26 above)
Comment (antonioT90) [reply src/test/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImplTest.java]: ok, let's rename it in order to give an hint on its purpose

File: src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java [hunk 30]
```diff
@@ -132,8 +140,12 @@ private TransactionResponse createTrxSuccess(TransactionCreationRequest trxReque
         return trxCreated;
     }
 
-    private void checkTransactionStored(TransactionResponse trxCreated) {
+    private void checkTransactionStored(TransactionResponse trxCreated) throws Exception {
         TransactionInProgress stored = checkIfStored(trxCreated.getId());
+        SyncTrxStatusDTO storedMapped = new TransactionInProgress2SyncTrxStatusMapper().transactionInProgressMapper(stored);
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java:145]: store the mapper as an instance variable instead of creating it for each invocation!
Moreover, it could be more readable to have the mapper invocation inside the asserEquals itself, as done in row 150

File: src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java [hunk 31]
```diff
@@ -0,0 +1,13 @@
...
+
+import it.gov.pagopa.payment.dto.mapper.TransactionInProgress2SyncTrxStatusMapper;
+import it.gov.pagopa.payment.dto.qrcode.SyncTrxStatusDTO;
+import it.gov.pagopa.payment.enums.SyncTrxStatus;
+import it.gov.pagopa.payment.model.TransactionInProgress;
+
+public class SyncTrxStatusFaker {
+    public static SyncTrxStatusDTO mockInstance(Integer bias, SyncTrxStatus status){
+    TransactionInProgress trx= TransactionInProgressFaker.mockInstance(bias, status);
+        return new TransactionInProgress2SyncTrxStatusMapper().transactionInProgressMapper(trx);
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java:11]: store the mapper as an instance variable instead of creating it each time

When an object represent a stateless re-usable instance, we could think to save some memory and process effort creating it just once

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java [hunk 32]
```diff
@@ -132,8 +142,11 @@ private TransactionResponse createTrxSuccess(TransactionCreationRequest trxReque
         return trxCreated;
     }
 
-    private void checkTransactionStored(TransactionResponse trxCreated) {
+    private void checkTransactionStored(TransactionResponse trxCreated) throws Exception {
         TransactionInProgress stored = checkIfStored(trxCreated.getId());
+        SyncTrxStatusDTO syncTrxStatusResult =extractResponse(getStatusTransaction(trxCreated.getId(), trxCreated.getMerchantId(), trxCreated.getAcquirerId()),HttpStatus.OK, SyncTrxStatusDTO.class);
+        extractResponse(getStatusTransaction(trxCreated.getId(), "DUMMYMERCHANTID", trxCreated.getAcquirerId()),HttpStatus.NOT_FOUND, null);
+        Assertions.assertEquals(transactionInProgress2SyncTrxStatusMapper.transactionInProgressMapper(stored),syncTrxStatusResult);
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/controller/BasePaymentControllerIntegrationTest.java:149]: reverse these 2 instructions in order to complete the 200 test before to test the 404

File: src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java [hunk 33]
```diff
@@ -0,0 +1,15 @@
+package it.gov.pagopa.payment.test.fakers;
+
+import it.gov.pagopa.payment.dto.mapper.TransactionInProgress2SyncTrxStatusMapper;
+import it.gov.pagopa.payment.dto.qrcode.SyncTrxStatusDTO;
+import it.gov.pagopa.payment.enums.SyncTrxStatus;
+import it.gov.pagopa.payment.model.TransactionInProgress;
+
+public class SyncTrxStatusFaker {
+    private static TransactionInProgress2SyncTrxStatusMapper transactionMapper;
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/payment/test/fakers/SyncTrxStatusFaker.java]: you have to call the new instance in order to create it

**Review by antonioT90 — CHANGES_REQUESTED** [review]

**Review by antonioT90 — APPROVED** [review]
